packages/core/contracts/l2/polygon/OracleRootTunnel.sol

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./tunnel/FxBaseRootTunnel.sol";
+import "./OracleBaseTunnel.sol";
+import "../../oracle/interfaces/OracleAncillaryInterface.sol";
+
+/**
+ * @title Adapter deployed on mainnet that validates and sends price requests from sidechain to the DVM on mainnet.
+ * @dev This contract must be a registered financial contract in order to make DVM price requests.
+ */
+contract OracleRootTunnel is OracleBaseTunnel, FxBaseRootTunnel {
+    constructor(
+        address _checkpointManager,
+        address _fxRoot,
+        address _finderAddress
+    ) OracleBaseTunnel(_finderAddress) FxBaseRootTunnel(_checkpointManager, _fxRoot) {}
+
+    /**
+     * @notice This is the first method that should be called in order to publish a price request to the sidechain.
+     * @dev Publishes the DVM resolved price for the price request, or reverts if not resolved yet.
+     * @param identifier Identifier of price request to resolve.
+     * @param time Timestamp of price request to resolve.
+     * @param ancillaryData extra data of price request to resolve.
+     */
+    function publishPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public {
+        require(_getOracle().hasPrice(identifier, time, ancillaryData), "DVM has not resolved price");
+        int256 price = _getOracle().getPrice(identifier, time, ancillaryData);
+        // This implementation allows duplicate MessageSent events via _sendMessageToRoot. The child tunnel on the
+        // sidechain will not have a problem handling duplicate price resolutions (it will just ignore them). This is
+        // potentially a fallback in case the automatic state sync to the sidechain is missing the `publishPrice`
+        // transaction for some reason. There is little risk in duplicating MessageSent emissions because the sidechain
+        // bridge does not impose any rate-limiting.
+        _publishPrice(identifier, time, ancillaryData, price);
+        // Initiate cross-chain price request:
+        // TODO: Should we pack more information into this request?
+        _sendMessageToChild(abi.encode(identifier, time, ancillaryData, price));
+    }
```

Comment:
are there any events you would like to emit over and above that produced by `_publishPrice`?

Reply:
I don't see anything, do you?

Reply:
I dont think so. I think you hit it just wanted to think it through fully as always more painful to add them back later!